### PR TITLE
test: Unswizzle swizzled methods

### LINKED ISSUE #1207: test: Unswizzle swizzled methods
When calling `SentrySDK.close`, all swizzled methods could be unswizzled. We could keep a cache of the original method IMPs and use `method_setImplementation` to change the swizzled method back to its original. After doing so, we should also call `imp_removeBlock` and remove the IMP from the cache.

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Added a way to unzwizzle methods.

## :bulb: Motivation and Context

closes #1207 

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [ ] Review from the native team if needed
- [x] No breaking changes

## :crystal_ball: Next steps

_#skip-changelog_

### CODE REVIEW

File: Sources/Sentry/SentrySwizzle.m [hunk 1]
```diff
@@ -102,11 +104,59 @@ @implementation SentrySwizzle
 
     pthread_mutex_lock(&gLock);
 
+#if TEST
+    IMP currImp = class_getMethodImplementation(classToSwizzle, selector);
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySwizzle.m:108]: `m`: Can't we use the `originalIMP` instead of this?

File: Sources/Sentry/SentrySwizzle.m [hunk 2]
```diff
@@ -102,11 +104,59 @@ @implementation SentrySwizzle
...
     originalIMP = class_replaceMethod(classToSwizzle, selector, newIMP, methodType);
 
     pthread_mutex_unlock(&gLock);
 }
 
+static void
+unswizzle(Class classToSwizzle, SEL selector)
+{
+#if TEST
+    static pthread_mutex_t gLock = PTHREAD_MUTEX_INITIALIZER;
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySwizzle.m:124]: `m`: This code is copied from https://github.com/google/GoogleUtilities/blob/797005ad8a1f0614063933e2fa010a5d13cb09d0/GoogleUtilities/SwizzlerTestHelpers/GULSwizzler%2BUnswizzle.m. Why don't we also copy this class?

File: Sources/Sentry/SentrySwizzle.m [hunk 3]
```diff
@@ -102,11 +104,59 @@ @implementation SentrySwizzle
...
+                                                          aClass:classToSwizzle];
+
+    pthread_mutex_unlock(&gLock);
+
+#endif
+}
+
++ (void)resetSwizzling
+{
+    NSArray *cached = [[GULSwizzlingCache sharedInstance] cachedClasses];
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySwizzle.m:149]: `h`: As `GULSwizzlingCache` is not thread-safe we need to surround this code with a lock I think.

File: Sources/Sentry/GULSwizzlingCache.m [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    const void *classSELCArray[2] = { (__bridge void *)(aClass), selector };
+    CFArrayRef classSELPair = CFArrayCreate(kCFAllocatorDefault, classSELCArray,
+        2, // Size.
+        NULL); // Elements are pointers so this is NULL.
+    CFDictionaryAddValue(_originalImps, classSELPair, originalIMP);
+
+    CFRelease(classSELPair);
+}
+
+- (NSArray *)cachedClasses
```
Comment (philipphofmann) [inline Sources/Sentry/GULSwizzlingCache.m:83]: `m`: Can we please specify the type here to know what the array contains?

File: Sources/Sentry/SentrySwizzle.m [hunk 5]
```diff
@@ -102,11 +104,59 @@ @implementation SentrySwizzle
...
+    for (int i = 0; i < cached.count; i++) {
+        CFArrayRef item = (__bridge CFArrayRef)cached[i];
+        Class class = (Class)CFArrayGetValueAtIndex(item, 0);
+        SEL selector = (SEL)CFArrayGetValueAtIndex(item, 1);
+        unswizzle(class, selector);
+    }
+
+    [swizzledClassesDictionary() removeAllObjects];
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySwizzle.m:157]: `h`: swizzledClassesDictionary is not thread safe and all the other usages use a `@synchronized` block.
```suggestion
    @synchronized(swizzledClassesDictionary()) {
        [swizzledClassesDictionary() removeAllObjects];
    }
```

File: Sources/Sentry/SentrySwizzle.m [hunk 6]
```diff
@@ -166,6 +216,28 @@ + (BOOL)swizzleInstanceMethod:(SEL)selector
...
+                return NO;
+            }
+        }
+    }
+
+    unswizzle(classToSwizzle, selector);
+
+    if (key) {
+        [swizzledClassesForKey(key) removeObject:classToSwizzle];
+    }
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySwizzle.m:236]: `h`: The synchronized block needs to include this piece of code. 
```suggestion
        }
        
        unswizzle(classToSwizzle, selector);
        
        if (key) {
            [swizzledClassesForKey(key) removeObject:classToSwizzle];
        }
    }
```

**Review by philipphofmann — COMMENTED** [review]
I know it's a draft, but I already did a thorough review anyways. Can we make sure that we don't ship the unswizzle code by using `#if Test` for everything?

File: Sources/Sentry/GULSwizzlingCache.m [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+{
+    NSDictionary *imps = (__bridge NSDictionary *)_originalImps;
+
+    // CFArrayRef array = imps.allKeys;
+    // Class class = CFArrayGetValueAtIndex(array, 0);
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/GULSwizzlingCache.m:88]: Oh, some test left behind. Thanks

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 ## Unreleased
 
+- feat: Add a way to reverse swizzling (#1593)
```
[the remark this answers is not in the document]
Comment (brustolin) [reply CHANGELOG.md:5]: Ok. Make sense

File: Sources/Sentry/include/SentrySwizzle.h [hunk 9]
```diff
@@ -287,6 +287,17 @@ typedef NS_ENUM(NSUInteger, SentrySwizzleMode) {
...
+/**
+ Revert any previously swizzling done in the class for the selector, and cleans the the swizzling
+ key indicator.
+ */
++ (BOOL)unswizzleInstanceMethod:(SEL)selector inClass:(Class)classToSwizzle key:(const void *)key;
+
+/**
+ Revert all previously swizzling.
+ */
++ (void)resetSwizzling;
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/include/SentrySwizzle.h:299]: I didn't like `unswizzleEverything` as a method name 😅. I'll came up with something else.

File: Sources/Sentry/SentrySwizzle.m (hunk 2 above)
Comment (brustolin) [reply Sources/Sentry/SentrySwizzle.m:124]: It's not really a class, but an extensions of the `GULSwizzler`, which we don't use. Coping just a few lines sees easier. However I could include the reference.

File: Sources/Sentry/SentrySwizzle.m (hunk 1 above)
Comment (brustolin) [reply Sources/Sentry/SentrySwizzle.m:108]: No. `originalIMP` may be null if the class don't have a direct implementation of it (inheritance).

File: Sources/Sentry/GULSwizzlingCache.m (hunk 4 above)
Comment (philipphofmann) [reply Sources/Sentry/GULSwizzlingCache.m:84]: I think this is not resolved yet. The header specifies `NSArray<NSArray *> *`, but can we also update this here and be more specific?

**Review by philipphofmann — COMMENTED** [review]
I think we can make this PR ready for review, right?

**Comment by brustolin** [comment]
> I think this is not resolved yet. The header specifies NSArray<NSArray *> *, but can we also update this here and be more specific?

I will update the implementation to reflect the header, but this inner array contains a `Class` and a `SEL`, so we cannot be more specific here.

**Review by philipphofmann — APPROVED** [review]
LGTM

File: Tests/SentryTests/Performance/SentryTracerTests.swift [hunk 10]
```diff
@@ -242,6 +242,12 @@ class SentryTracerTests: XCTestCase {
         let child2 = sut.startChild(operation: fixture.transactionOperation)
         let child3 = sut.startChild(operation: fixture.transactionOperation)
         child2.finish()
+        
+        //Without this sleep sut.timestamp and child2.timestamp sometimes
+        //are equal. We need to make sure that SentryTracer is not changing
+        //the timestamp value of proper finished spans.
+        Thread.sleep(forTimeInterval: 0.1)
```
Comment (philipphofmann) [inline Tests/SentryTests/Performance/SentryTracerTests.swift:249]: `m`: Can't we achieve the same by changing the time of the dateprovider instead with something like `fixture.dateProvider.setDate(date: fixture.dateProvider.date().addingTimeInterval(bySeconds))`?

Comment (philipphofmann) [reply Tests/SentryTests/Performance/SentryTracerTests.swift:249]: Ah now I get it thanks.

**Comment by philipphofmann** [comment]
Closed, because we had several problems we could not solve in a reasonable amount of time when trying unswizzling. As swizzling opens up the possibility for new hard-to-track-down bugs, we only wanted to use unswizzling in tests. The main argument for this task was the hypothesis that tests would get more reliable. Instead, we experience flaky tests like crazy. The tests crash in CI mostly with SIGSEV, and it's hard to reproduce this behavior locally. Therefore the feedback loop is sluggish when trying to hunt down the root cause of the problem. While investigating, we discovered that the crash logs of tests can be found in the test logs. We opened a PR for this https://github.com/getsentry/sentry-cocoa/pull/1621, so if we have similar problems in the future, the investigation is more straightforward.
Maybe our approach with using the [GULSwizzlingCache](https://github.com/google/GoogleUtilities/blob/797005ad8a1f0614063933e2fa010a5d13cb09d0/GoogleUtilities/SwizzlerTestHelpers/GULSwizzlingCache.m) from is not the right one because our way of swizzling is different that from the Google Utilities. It could be that we make it work when trying a different approach, but as we don't dare to ship unswizzling in the SDK and only use it for tests we don't think it's worth the effort. Instead, our code should make sure that our swizzling always calls the original implementation, which it already does. When closing the SDK, the code being called from swizzling shouldn't do anything, which we also already do.

We is @philipphofmann  and @brustolin.

**Pull request closed without merging**